Add support for symlink kernel caching

HereThereBeDragons opened the pull request:
Extended @mharvey-jt PR #2959 :
- fixes the wrongly truncated new symlinks when short -> long
- added fallback in case symlink caching is not possible 
- and fixed some small things.

Note: If you use this PR, enabling symlink kernel caching, you SHOULD NOT use symlinks as mountpoint. 
Currently it is not possible to just invalidate dentries in away that they are just considered stale in cache and need revalidation. Instead, the dentry of the symlink is evicted and the mountpoint on the symlink will be just gone.

The current libfuse implementation does not offer any solution to this problem.

Missing: Tests..

Fixes #2949

mharvey-jt committed:
CVMFS_CACHE_SYMLINKS

HereThereBeDragons committed:
added fallback in case symlink caching is not possible

HereThereBeDragons committed:
add debug output for failing fuse_inval calls

HereThereBeDragons committed:
fix cpplint complaints

jblomer commented:
Related discussion on the fuse mailing list: https://sourceforge.net/p/fuse/mailman/message/37702061/

HereThereBeDragons committed:
WIP - tests working but need cleanup

HereThereBeDragons committed:
cleanup+comments done

HereThereBeDragons committed:
dentry eviction is now always called if symlink caching is activated, update in logging. might want to reverse that at a later point

HereThereBeDragons committed:
WIP server/client integration test including mount_on_top for symlink kernel caching

HereThereBeDragons committed:
Integration Test 702: Cleanup finished

HereThereBeDragons committed:
appease linter

HereThereBeDragons committed:
READ COMMIT! Adds support to expire dentries

Prevents mount-on-top mountpoints breaking and fixes symlink kernel caching

NOTE: libfuse patch provides fuse_lowlevel_notify_expire_entry()
BUT IT DOES NOT PROVIDE: FUSE_CAP_DENTRY_EXPIRE_ONLY
This needs to be defined yourself in libfuse/include/fuse_common.h
e.g. like this: #define FUSE_CAP_DENTRY_EXPIRE_ONLY (1 << 26)

HereThereBeDragons committed:
appease linter

HereThereBeDragons commented:
Working tests for kernel caching including symlinks for fuse2 and >= fuse3.10. Tested on fuse3.10 with kernel patch and without
- [patched_kernel_702-symlink_caching.log](https://github.com/cvmfs/cvmfs/files/9617655/patched_kernel_702-symlink_caching.log)
- [unpatched_kernel_702-symlink_caching.log](https://github.com/cvmfs/cvmfs/files/9617656/unpatched_kernel_702-symlink_caching.log)

**Thinks that need further decision**

0) Ask libfuse to add some CAP_ define to check with the kernel supports notify_expire_entry
  - libfuse patch provides fuse_lowlevel_notify_expire_entry()
  - BUT in addition, i introduced: FUSE_CAP_DENTRY_EXPIRE_ONLY in libfuse/include/fuse_common.h to be able to see if the kernel supports it

1) One problem of the test is the clean up afterwards. It needs sudo rights
  ```
  -- Testing Symlink caching: Proper eviction policy for kernel cache (Wed Sep 21 16:23:48 CEST 2022 / test number 702)
  rm: cannot remove '/tmp/cvmfs-test/workdir/702-symlink_caching/private_mntc/just.test.repo': Permission denied
  ```

2) We should think about these lines 469-472 in `cvmfs.cc`:
```
  if (mount_point_->cache_symlinks()) {  //  && dirent.IsLink()) {
    LogCvmfs(kLogCache, kLogDebug, "Dentry to evict: %s", name);
    mount_point_->dentry_tracker()->Add(parent_fuse, name, uint64_t(timeout));
  }
```

we can do something like
```c++
#ifdef FUSE_CAP_DENTRY_EXPIRE_ONLY //we can always deleten dentry tracker
    mount_point_->dentry_tracker()->Add(parent_fuse, name, uint64_t(timeout));
#else // we only delete if symlink caching is requested and only symlink dentries
  if (mount_point_->cache_symlinks())  && dirent.IsLink()) {
    LogCvmfs(kLogCache, kLogDebug, "Dentry to evict: %s", name);
    mount_point_->dentry_tracker()->Add(parent_fuse, name, uint64_t(timeout));
  }
#endif
```

Questions:
- Keep log message? (I just had it for testing) `LogCvmfs(kLogCache, kLogDebug, "Dentry to evict: %s", name);`

@jblomer (just to notify you)

HereThereBeDragons commented:
Ah and another thing: the linter here on github currently returns the wrong line numbers? 
If you call it locally you get the correct line numbers back

jblomer commented:
> Ah and another thing: the linter here on github currently returns the wrong line numbers? If you call it locally you get the correct line numbers back

That should eventually be fixed by [moving to clang-format](https://sft.its.cern.ch/jira/browse/CVM-2030).

jblomer commented:
>     0. Ask libfuse to add some CAP_ define to check with the kernel supports notify_expire_entry
> 
> 
>     * libfuse patch provides fuse_lowlevel_notify_expire_entry()
> 
>     * BUT in addition, i introduced: FUSE_CAP_DENTRY_EXPIRE_ONLY in libfuse/include/fuse_common.h to be able to see if the kernel supports it

Yep, that should be part of our feedback to the fuse mailing list.


>     1. One problem of the test is the clean up afterwards. It needs sudo rights
> 
> 
> ```
> -- Testing Symlink caching: Proper eviction policy for kernel cache (Wed Sep 21 16:23:48 CEST 2022 / test number 702)
> rm: cannot remove '/tmp/cvmfs-test/workdir/702-symlink_caching/private_mntc/just.test.repo': Permission denied
> ```

That's fine, we can just use `sudo` where needed. Integration tests are supposed to run on an account that has password-less sudo access.

>     2. We should think about these lines 469-472 in `cvmfs.cc`:
> 
> 
> ```
>   if (mount_point_->cache_symlinks()) {  //  && dirent.IsLink()) {
>     LogCvmfs(kLogCache, kLogDebug, "Dentry to evict: %s", name);
>     mount_point_->dentry_tracker()->Add(parent_fuse, name, uint64_t(timeout));
>   }
> ```
> 
> we can do something like
> 
> ```c++
> #ifdef FUSE_CAP_DENTRY_EXPIRE_ONLY //we can always deleten dentry tracker
>     mount_point_->dentry_tracker()->Add(parent_fuse, name, uint64_t(timeout));
> #else // we only delete if symlink caching is requested and only symlink dentries
>   if (mount_point_->cache_symlinks())  && dirent.IsLink()) {
>     LogCvmfs(kLogCache, kLogDebug, "Dentry to evict: %s", name);
>     mount_point_->dentry_tracker()->Add(parent_fuse, name, uint64_t(timeout));
>   }
> #endif
> ```

Hm, `ifdef` may not be enough because both kernel and library need to support it.  We need to check dynamically if the capability is set (see `fuse_conn_info->capable` and `fuse_conn_info->want`).

I think we should by default disable symlink caching if expiring dentries is not available, even if enable by cvmfs options. In this case, we should add a warning to syslog ("symlink caching enabled but not available..."). We may consider an additional option that enforces symlink caching even if the fuse requirements are missing.  In this case, we would resort to only add symlinks to the dentry tracker (and warn about breaking mountpoints on symlinks).

> Questions:
> 
>     * Keep log message? (I just had it for testing) `LogCvmfs(kLogCache, kLogDebug, "Dentry to evict: %s", name);`

This can stay, I think.

jblomer commented:
On second thought, regarding cvmfs options: we actually don't need an option `CVMFS_SYMLINK_CACHING`, I think. If fuse supports symlink caching _and_ dentry expiry, we just enable it.  Otherwise we don't. We may have an option `CVMFS_ENFORCE_SYMLINK_CACHING` for a transition period, where we enable symlink caching (with caveats) even if we cannot expire dentries.  Let's merge a first PR without that option though.

HereThereBeDragons commented:
ok then we first need a new version of fuse where we can check if the capability is available.
(thought it might be easier to commit immediately the `CVMFS_ENFORCE_SYMLINK_CACHING` because we already have all the functionality of it in the previous PR)

and i guess you are right that a "define" is not good enough as its defined in libfuse and not in the fuse kernel module

HereThereBeDragons committed:
updating patch to support libfuse patch independent of fuse version (backport possibility)

HereThereBeDragons committed:
appeasing linter

HereThereBeDragons commented:
@jblomer have a look. 
not sure how we can test it automatically unless we patch our own kernel for some of the cloudtest machines..

jblomer commented on test/test_functions:
```diff
@@ -2372,10 +2374,14 @@ EOF
   echo "  *** local mount config:"
   cat "$config"
 
+  if [[ ! -z $mount_options ]]; then
```
`[[ ]]` is a bashism (doesn't work on Ubuntu's dash)
```suggestion
  if [ ! -z $mount_options ]; then
```

jblomer commented on test/test_functions:
```diff
@@ -2372,10 +2374,14 @@ EOF
   cat "$config"
 
+  if [[ ! -z $mount_options ]]; then
+    mount_options=$(echo $mount_options",")
```
```suggestion
    mount_options="${mount_options},"
```

jblomer commented on cvmfs/fuse_evict.cc:
```diff
@@ -190,14 +199,26 @@ void *FuseInvalidator::MainInvalidator(void *data) {
+#ifdef FUSE_CAP_EXPIRE_ONLY
+      fuse_lowlevel_notify_expire_entry(channel_or_session,
+        entry_parent, entry_name.GetChars(), entry_name.GetLength(),
+        fuse_expire_flags::FUSE_LL_EXPIRE_ONLY);
```
I guess this doesn't work with C++03
```suggestion
        FUSE_LL_EXPIRE_ONLY);
```

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1633,6 +1642,34 @@ static void cvmfs_init(void *userdata, struct fuse_conn_info *conn) {
+                                    "Enable symlink caching");
+    } else {
+      mount_point_->DisableCacheSymlinks();
+      LogCvmfs(kLogCvmfs, kLogDebug | kLogSyslogErr,
```
```suggestion
      LogCvmfs(kLogCvmfs, kLogDebug | kLogSyslogWarn,
```

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1633,6 +1642,34 @@ static void cvmfs_init(void *userdata, struct fuse_conn_info *conn) {
+    }
+#else
+    mount_point_->DisableCacheSymlinks();
+    LogCvmfs(kLogCvmfs, kLogDebug | kLogSyslogErr,
```
```suggestion
    LogCvmfs(kLogCvmfs, kLogDebug | kLogSyslogWarn,
```

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1633,6 +1642,34 @@ static void cvmfs_init(void *userdata, struct fuse_conn_info *conn) {
+    LogCvmfs(kLogCvmfs, kLogDebug, "FUSE: "
+                                   "Enable fuse_expire_entry");
+  }
+#endif
```
I think we want an `#else` branch and print one more warning: if we have symlink caching turned on but expire entry turned off, we want to issue a warning about the fact that mountpoints on top of symlinks will break.

jblomer commented on test/src/702-symlink_caching/main:
```diff
@@ -0,0 +1,196 @@
+  trap cleanup_mount_on_top EXIT HUP INT TERM
+
+  ls $mntpnt/testdir_mountpoint/self > /dev/null || return 34
+  diff $mntpnt/testdir_mountpoint/self/maps /proc/self/maps || return 35
```
I guess this fails on macOS, so perhaps restrict this check to linux only; or remove.

HereThereBeDragons replied:
why? other directory possible to mount on top that exists on all platforms?

jblomer commented on test/src/702-symlink_caching/main:
```diff
@@ -0,0 +1,196 @@
+  replace_short_symlink_with_long $short_name $long_name $symlink_name
+  local new_symlink_pointer=$(readlink ${mntpnt}/${symlink_name})
+  # res_switch must be 0
+  if [ "$new_symlink_pointer" = "$long_name" ] && [ "$new_symlink_pointer" != "$short_name" ]; then
```
```suggestion
  if [ "$new_symlink_pointer" = "$long_name" ] && [ "$old_symlink_pointer" = "$short_name" ]; then
```

jblomer reviewed (changes requested):
Very nice!  Almost there, I think.  We'll need to exclude the 702 integration test on a number (all?) of platforms.

jblomer commented:
@cernvm-bot cloudtest

jblomer commented:
@cernvm-bot cloudtest

jblomer commented:
@cernvm-bot cloudtest

HereThereBeDragons committed:
nearly all change requests but "diff ... /proc/self/maps" in test

HereThereBeDragons commented:
> Very nice! Almost there, I think. We'll need to exclude the 702 integration test on a number (all?) of platforms.

i removed it now from all platforms. can be added back if better support for symlink caching and expire_entries is given. 

if backporting to libfuse2 becomes a common thing the integration test needs to be modified

HereThereBeDragons committed:
appease clang tidy

HereThereBeDragons commented:
@cernvm-bot cloudtest

HereThereBeDragons commented:
@jblomer  have a look. seems like the test fails do to unrelated stuff
